### PR TITLE
fix: Log auth events for OIDC login failure and success

### PR DESCRIPTION
**Pull request by netroms**
OIDC login attempts, both successful and unsuccessful are not logged in the same way as password logins. 

Signed-off-by: Morten Svanaes <msvanaes@dhis2.org>

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java [hunk 1]
```diff
@@ -46,42 +53,75 @@
...
-            final StringBuilder builder = new StringBuilder();
-            builder.append( "Authentication event " );
-            builder.append( ClassUtils.getShortName( event.getClass() ) );
-            builder.append( ": " );
-            builder.append( event.getAuthentication().getName() );
+            String authName = event.getAuthentication().getName();
+            String eventClassName = "Authentication event: " + ClassUtils.getShortName( event.getClass() );
 
-            Object details = event.getAuthentication().getDetails();
+            String exceptionMessage = "";
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java:67]: Newline

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java [hunk 2]
```diff
@@ -46,42 +53,75 @@
...
+            if ( event instanceof AbstractAuthenticationFailureEvent )
+            {
+                exceptionMessage =
+                    "; exception: " + ((AbstractAuthenticationFailureEvent) event).getException().getMessage();
+            }
+
+            String ipAddress = "";
+            String sessionId = "";
 
+            Object details = event.getAuthentication().getDetails();
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java:77]: Newline

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java [hunk 3]
```diff
@@ -46,42 +53,75 @@
...
+                sessionId = hashSessionId( authDetails.getSessionId() );
+            }
+            else if ( OAuth2LoginAuthenticationToken.class.isAssignableFrom( event.getAuthentication().getClass() ) )
+            {
+                OAuth2LoginAuthenticationToken authenticationToken = (OAuth2LoginAuthenticationToken) event
+                    .getAuthentication();
 
-                String sessionId = authDetails.getSessionId();
-                if ( sessionId != null )
+                DhisOidcUser principal = (DhisOidcUser) authenticationToken.getPrincipal();
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java:90]: Newline

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java [hunk 4]
```diff
@@ -46,42 +53,75 @@
...
     }
+
+    private String hashSessionId( String sessionId )
+    {
+        if ( sessionId == null )
+        {
+            return "";
+        }
+
+        HashCode hash = Hashing.sha256().newHasher().putString( sessionId, Charsets.UTF_8 ).hash();
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java:124]: Does it make sense to cache this hasher? or even just use `Hashing.sha256().hashString(sessionId, StandardCharsets.UTF_8).toString()` direcly?

File: dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/security/config/AuthenticationListener.java [hunk 5]
```diff
@@ -66,33 +68,67 @@
...
+                (TwoFactorWebAuthenticationDetails) auth.getDetails();
 
             log.info( String.format( "Login attempt failed for remote IP: %s", authDetails.getIp() ) );
         }
 
-        securityService.registerFailedLogin( auth.getName() );
+        if ( OAuth2LoginAuthenticationToken.class.isAssignableFrom( auth.getClass() ) )
+        {
+            OAuth2LoginAuthenticationToken authenticationToken = (OAuth2LoginAuthenticationToken) auth;
+            DhisOidcUser principal = (DhisOidcUser) authenticationToken.getPrincipal();
```
Comment (mortenoh) [inline dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/security/config/AuthenticationListener.java:84]: Newline

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java (hunk 4 above)
Comment (netroms) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/AuthenticationLoggerListener.java:124]: I made the HashingFunction a field. The rest of the API is mostly immutable so object reuse is not possible.

**Review by mortenoh — APPROVED** [review]

**Review by stian-sandvold — APPROVED** [review]

**Pull request merged**